### PR TITLE
[std/locks]close #7998(complete condition variables)

### LINKED ISSUE #7998: The condition variables API is incomplete
At the moment Nim offers just a pair of `signal` and `wait` APIs for its condition variables:
https://github.com/nim-lang/Nim/blob/master/lib/core/locks.nim#L5

In other programming languages there are separate calls such as `notifyOne`, `notifyAll` or even `notify(n)` when signalling the condition:

https://en.cppreference.com/w/cpp/thread/condition_variable
https://docs.python.org/3/library/asyncio-sync.html#asyncio.Condition

Furthermore, perhaps async options may be provided over the same underlying OS primitives. @cheatfate should be able to comment on this as he has implemented similar APIs in his [asynctools](https://github.com/cheatfate/asynctools) repo.

The [threadpool module](https://github.com/nim-lang/Nim/blob/devel/lib/pure/concurrency/threadpool.nim#L20) defines additional primives such as a `Semaphore` which are currently not exported. I think these can be moved to the `locks` module and exported.

### LINKED ISSUE #17669: conditional broadcast missing from locks module
### Summary

`broadcast` is used to wake up blocked threads that are placed in a queue. `signal` on the other hand unblocks a single thread. Some applications require that all threads are [awaken](
http://www.qnx.com/developers/docs/qnxcar2/topic/com.qnx.doc.neutrino.getting_started/topic/s1_procs_Condvar_signal_vs_broadcast.html) . This api is currently missing.

### PR DESCRIPTION
**Pull request by ringabout**
Ref https://github.com/timotheecour/Nim/issues/693

close #7998
close https://github.com/nim-lang/Nim/issues/17669

`genode` OS is still left unchanged, I have no idea.

After CI is green, I will add changelog, tests and since.

It is the basis of concurrency programming.

In the future, `Semaphore` should be exported or re-implemented.

### CODE REVIEW

**Comment by ringabout** [comment]
This PR workarounds the broadcast problem on genode, it is fixable in the future PR.

Ref: https://www.microsoft.com/en-us/research/wp-content/uploads/2004/12/ImplementingCVs.pdf

**Comment by ringabout** [comment]
Not sure why it doesn't work, maybe it is due to spawn is compiler magics(boost issue)?

**Comment by ringabout** [comment]
Maybe we need a modern lock/cond library with destructors. In the future add `SpinLock`, `RwLock`, `Semaphore`, `Barrier`, `Once` or so.

https://gist.github.com/xflywind/eb13d4712ec5e0a3fd3d09501b813831


At least now we need `std/nlocks` + `std/conds`.

What's your opinion @timotheecour ?

File: lib/system/syslocks.nim [hunk 1]
```diff
@@ -48,30 +50,36 @@ when defined(windows):
...
 
-  proc signalSysCond(hEvent: SysCond) {.stdcall, noSideEffect,
-    dynlib: "kernel32", importc: "SetEvent".}
+  proc wakeConditionVariable(
+    conditionVariable: var SysCond
+  ) {.stdcall, noSideEffect, dynlib: "kernel32", importc: "WakeConditionVariable".}
+
+  proc wakeAllConditionVariable(
+    conditionVariable: var SysCond
+  ) {.stdcall, noSideEffect, dynlib: "kernel32", importc: "wakeAllConditionVariable".}
```
Comment (ringabout) [inline lib/system/syslocks.nim]:
```suggestion
```

File: lib/system/syslocks.nim [hunk 2]
```diff
@@ -48,30 +50,29 @@ when defined(windows):
...
+    conditionVariable: var SysCond
+  ) {.stdcall, noSideEffect, dynlib: "kernel32", importc: "InitializeConditionVariable".}
 
-  proc closeHandle(hObject: Handle) {.stdcall, noSideEffect,
-    dynlib: "kernel32", importc: "CloseHandle".}
-  proc waitForSingleObject(hHandle: Handle, dwMilliseconds: int32): int32 {.
-    stdcall, dynlib: "kernel32", importc: "WaitForSingleObject", noSideEffect.}
+  proc sleepConditionVariableCS(
+    conditionVariable: var SysCond,
+    PCRITICAL_SECTION: SysLock,
```
Comment (ringabout) [inline lib/system/syslocks.nim]:
```suggestion
    PCRITICAL_SECTION: var SysLock,
```

**Comment by Araq** [comment]
I'm not convinced that "broadcast" is a good design. If the threads check for different conditions then all waked up threads would cause contention on the condition variable's lock afterwards...  The fact that its support needed these signification Windows API wrapper changes doesn't make it better. What are the use cases for "broadcast"? I have yet to see a convincing one.

**Comment by ringabout** [comment]
For example implementing a barrier: when `cond` is ready, you need to unblock all threads instead of one or two threads. 

https://github.com/boostorg/thread/blob/4abafccff4bdeb4b5ac516ff0c2bc7c0dad8bafb/include/boost/thread/barrier.hpp#L229

The Nim one:
https://github.com/planetis-m/nim-100days/blob/adee939211ecfadb138b4da4a2209c4340ece2da/multithread/pbarrier2.nim#L30

https://stackoverflow.com/questions/9815798/how-to-find-barrier-functions-implementation

> cause contention on the condition variable's lock afterwards.

that's true, but we have to wake up all thread.

see http://www.qnx.com/developers/docs/qnxcar2/index.jsp?topic=%2Fcom.qnx.doc.neutrino.getting_started%2Ftopic%2Fs1_procs_Condvar_signal_vs_broadcast.html


So the usecase
> The threads are all unique and are each waiting for a very specific condition to occur.

> This nicely reflects the second case in our question above ("Why are these threads waiting?"). Since the threads are all waiting on different conditions (thread1() is waiting for x to be less than or equal to 7 or y to be 15, thread2() is waiting for x to be a prime number, and thread3() is waiting for x to be equal to y), we have no choice but to wake them all.


So if tasks are the same, `signal` should be used; while they are unique, `broadcast` should be used.

**Comment by saem** [comment]
`broadcast` provides an optimization opportunity for the underlying threading implementation that wouldn't be possible if one was to instead repeatedly `signal`. The additional benefit I didn't think of is that designs using broadcast encourage the handling of spurious wakes, which will also occur in situations where `signal` is used.

The spurious wake issue is documented here in the rather short rationale section ([https://pubs.opengroup.org/onlinepubs/009696699/functions/pthread_cond_broadcast.html])

If the intention with what Nim exposes is to provide more Nim specific semantics that apply across systems, as opposed to lightly wrapping, then I can see why we'd want to think deeply about including`broadcast`. If it is to provide a light wrapper for largely similar semantics and paper over API differences then I'd recommend including it.

**Comment by planetis-m** [comment]
At first I thought I could do without broadcast in the `Barrier` implementation, thinking each thread would wake the next thread, etc until the last one would maybe signal the thread that waits on the next barrier and cause a spurious wake or just fail to find another suspended thread.

However soon enough, I hit an edge case where some other thread than the last, would wake a thread that is not at the current barrier causing the above logic to fail.

 **Point is `signal` is not a replacement for `broadcast`**. Plus on a benchmark I did (with barriers), broadcast was twice as fast as signal (provided that it didn't get stuck).

**Comment by planetis-m** [comment]
Forgot to add that one could make a barrier implementation using [atomics](https://stackoverflow.com/questions/24205226/how-to-implement-a-re-usable-thread-barrier-with-stdatomic).

**Comment by Araq** [comment]
> However soon enough, I hit an edge case where some other thread than the last, would wake a thread that is not at the current barrier causing the above logic to fail.

How could this happen if every barrier had its own condition variable?

But ok, let's have "broadcast" then because all the others have it too, Turning the thundering herd problem into an API, what could go wrong.

**Comment by planetis-m** [comment]
> How could this happen if every barrier had its own condition variable?

Sorry to be more precise, I meant the local generation count, they are reusable barriers after all. It fits with the rule xflywind mentioned. [This](https://www.youtube.com/watch?v=xkpNYZIPTR4) is a better explanation than I can provide.

> Turning the thundering herd problem into an API, what could go wrong.

I believe linux solves this with ["wait morphing"](https://stackoverflow.com/questions/39252605/about-the-pthread-cond-wait), maybe this is a problem with windows.

**Comment by Araq** [comment]
> I believe linux solves this with "wait morphing", maybe this is a problem with windows.

Aha! Good to know!

**Pull request merged**